File: frameWork.py
# framework CRUD program

import json

MENUFILE = "menu.cfg"
FIELDFILE = "fields.cfg"
DATAFILE = "frameWorkRecords.dat"
DELETEDDATAFILE = "deletedRecords.dat"

def showMenu():
	while 1:
		printMenu()
		choice = int(input("Enter Your Choice: "))
		#operations = [create, showAll, update, delete, exit]
		#operations[choice - 1]()
		match choice:
			case 1:
				create()
			case 2:
				showAll()
			case 3:
				update()
			case 4:
				delete()
			case 5:
				print("Program Exited")
				exit(0)
			case _:
				print("Please Enter a Valid Choice!")

def create():
	readRecord()
	saveRecordsIntoFile()

def readRecord():
	"""for field in fields:
		fieldValue = input(f"Enter {field}: ")
		record.append(fieldValue)
	records.append(record)"""

	"""for fieldCounter in range(len(fields)):
		fieldValue = input(f"Enter {fields[fieldCounter]}: ")
		if (fieldCounter == (len(fields) - 1)):
			fieldValue = float(fieldValue)
		record.append(fieldValue)
	records.append(record)
	print("Saved Successfully!")"""

	record = []
	for counter, field in enumerate(fields):
		fieldValue = input(f"Enter {field}: ")
		if (counter == len(fields) - 1):
			fieldValue = float(fieldValue)
		record.append(fieldValue)
	records.append(record)
	print("Saved Successfully!")

def saveRecordsIntoFile():
	fpData = open(DATAFILE, "w")
	#fpData.write(str(records))
	json.dump(records, fpData)
	fpData.close()

def showAll():
	if (checkRecords()):
		return
	print("-----------")
	print("All records")
	print("-----------")
	for record in records:
		printRecord(record)

def printRecord(record):
	"""for record in records:  or #for recordCounter in range(len(records)):
		counter = 0
		for field in fields:
			#print(f"{field}: {records[recordCounter][counter]}")
			print(f"{field}: {record[counter]}")
			counter += 1 """
	for counter, field in enumerate(fields):
		print(f"{field}: {record[counter]}")

def update():
	if (checkRecords()):
		return
	tempRecord = searchRecord("Update")
	# print(id(tempRecord))
	if (tempRecord != None):
		tempRecord[len(fields) - 1] = float(input(f"Enter new {fields[len(fields) - 1]}: "))
		saveRecordsIntoFile()
		print(f"{fields[0]} {idToCheck} Updated Successfully!")
	else:
		print(f"{fields[0]} {idToCheck} is Not found!")

def searchRecord(operation):
	global idToCheck 
	idToCheck = input(f"Enter {fields[0]} to {operation}: ")
	for record in records:
		if (record[0] == idToCheck):
			#print(id(record))
			return record

def delete():
	if (checkRecords()):
		return
	tempRecord = searchRecord("Delete")
	if (tempRecord != None):
		#deletedRecords.append(tempRecord)
		saveDeletedIntoFile(tempRecord)
		records.remove(tempRecord)
		saveRecordsIntoFile()
		print(f"{fields[0]} {idToCheck} Deleted Successfully!")
	else:
		print(f"{fields[0]} {idToCheck} is Not Found!")

def saveDeletedIntoFile(tempRecord):
	fpDeleted = open(DELETEDDATAFILE, "a")
	fpDeleted.write(str(tempRecord))
	fpDeleted.close()

def printMenu():
	fpMenu = open(MENUFILE, "r")
	menu = fpMenu.read()
	if (not menu):
		print(f"{MENUFILE} is Empty!")
		fpMenu.close()
		return
	fpMenu.close()
	print(menu)

def loadFields():
	fpFields = open(FIELDFILE, "r")
	global fields 
	fields = fpFields.readlines()
	if (not fields):
		print(f"{FIELDFILE} is Empty!")
		fpFields.close()
		return 
	fpFields.close()
	for fieldCounter in range(len(fields) - 1):
		fields[fieldCounter] = fields[fieldCounter].strip()
	#print(fields)

def loadRecords():
	global records #without this it over writes in the empty globally declared variable
	fpData = open(DATAFILE, "r")
	content = fpData.read()
	# records = fpData.read() #reads as string. if readlines() is used then records are like list of string of list of list(["[[],[]]"])
	if (not content):
		records = []
		fpData.close()
		return
	records = json.loads(content)
	# records = eval(records) #changed to list of list
	fpData.close()
	# print(records) 

def checkRecords():
	if (not records):
		# print("No " + str(fields[0][:8]) + "Found!")
		print(f"No Records Found!") #records should be replaced by correct name like accounts, items, tickets. store it another .cfg and use it
		return 1

loadFields()
loadRecords()
showMenu()

File: frameWork.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define MENU_FILE "menu.cfg"
#define FIELD_FILE "fields.cfg"
#define DATA_FILE "records.dat"
#define LENGTH 20


void readMenu();
void save();
void update();
void delete();
void showAll();
void printMenu();
void loadFields();
void removeNewLine(char*);
void pause();

char *menu;
char **fieldNames;
int fieldCount;
char status;

int main() {
    
    	int choice;
	readMenu();
	loadFields();
   	do {
		system("cls");
		printMenu();
		scanf("%d", &choice);
		getchar();
		switch(choice)
		{
			case 1: save(); break;
			case 2: update(); break;
			case 3: delete(); break;
			case 4: showAll(); break;
			case 5: free(menu); 
				for(int fieldCounter = 0; fieldCounter < fieldCount; fieldCounter++)
				{
					free(fieldNames[fieldCounter]);
				}
				free(fieldNames);
				exit(0);
			default: printf("Invalid Choice!\n"); pause(); break;
		}
	} while(choice != 5);

    	return 0;
}

void readMenu()
{
    	FILE *fpData = fopen(MENU_FILE, "r");
    	if (!fpData) {
        	printf("File %s Not Found.\n", MENU_FILE);
        	return;
    	}
    	
	fseek(fpData, 0, SEEK_END);
	int sizeOfFile = ftell(fpData);
	//fseek(fpData, -sizeOfFile, SEEK_CUR);
	//fseek(fpData, 0, SEEK_SET);
	rewind(fpData);
	menu = malloc(sizeOfFile + 1);
	fread(menu, sizeOfFile , 1, fpData);
	menu[sizeOfFile] = '\0';
    	fclose(fpData);
}

void save()
{
       	FILE *fpFields = fopen(FIELD_FILE, "r");
    	if (!fpFields) {
        	printf("%s not found!\n", FIELD_FILE);
        	return;
    	}

        FILE *fpData = fopen(DATA_FILE, "a");
    	if (!fpData) {
        	printf("Cannot open %s\n", DATA_FILE);
        	fclose(fpFields);
        	return;
    	}
         
    	char fieldValue[LENGTH]; 
	 
    	printf("\nAccount Details\n------------\n");

        for(int fieldCounter = 0; fieldCounter < fieldCount - 1; fieldCounter++) {
	
        	fieldNames[strcspn(fieldNames[fieldCounter], "\r\n")] = '\0';

        	printf("Enter %s: ", fieldNames[fieldCounter]);
	
        	fgets(fieldValue, sizeof(fieldValue), stdin); 
		removeNewLine(fieldValue);

		char buffer[LENGTH] = {0};

            	strncpy(buffer, fieldValue, LENGTH);

            	fwrite(buffer, LENGTH, 1, fpData);
        	
    	}
	status = 'A';
	fwrite(&status, sizeof(status), 1, fpData);

    	fclose(fpFields);
    	fclose(fpData);

    	printf("\n Record saved successfully!\n");	
	pause();
}

void update()
{
	char fieldValueToCheck[LENGTH];
	char value[LENGTH];
	char newValue[LENGTH];
	int foundValue = 0;
	printf("Enter the %s: ", fieldNames[0]);
	fgets(fieldValueToCheck, LENGTH, stdin);
	removeNewLine(fieldValueToCheck);
	FILE *fpData = fopen(DATA_FILE, "r+");
	if (!fpData)
	{
		printf("File %s Not Found\n", DATA_FILE);
		return;
	}
	while(fread(value, LENGTH, 1, fpData))
	{
		removeNewLine(value);
		if(strcmp(fieldValueToCheck, value) == 0)
		{
			printf("Match Found!\n");
			printf("Enter new %s: ", fieldNames[fieldCount - 2]);
			fgets(newValue, LENGTH, stdin);
			removeNewLine(newValue);
			fseek(fpData, ((fieldCount - 2) * LENGTH) - LENGTH , SEEK_CUR);
					
			char buffer[LENGTH] = {0};
            		strncpy(buffer, newValue, LENGTH);
			fwrite(buffer, LENGTH, 1, fpData);
			foundValue = 1;
			printf("Updated successfully!\n");
			break;
		}
		else
		{
			fseek(fpData, ((fieldCount - 2) * LENGTH + 1), SEEK_CUR);
		}
		
	}
	if(!foundValue)
	{
		printf("%s Not Found!\n", fieldValueToCheck);
	}
	fclose(fpData);
	pause();
}

void delete()
{
	char fieldValueToCheck[LENGTH];
	char value[LENGTH];
	int foundValue = 0;
	char status;
	printf("Enter the %s: ", fieldNames[0]);
	fgets(fieldValueToCheck, LENGTH, stdin);
	removeNewLine(fieldValueToCheck);
	FILE *fpData = fopen(DATA_FILE, "r+");
	if (!fpData)
	{
		printf("File %s Not Found\n", DATA_FILE);
		return;
	}
	while(fread(value, LENGTH, 1, fpData))
	{
		removeNewLine(value);
		if(strcmp(fieldValueToCheck, value) == 0)
		{
			fseek(fpData, ((fieldCount - 2) * LENGTH), SEEK_CUR);
			status = 'D';
			fwrite(&status, sizeof(status), 1, fpData);
			foundValue = 1;
			printf("Deleted %s Successfully!\n", fieldValueToCheck);
			break;
		}
		else
		{
			fseek(fpData, ((fieldCount - 2) * LENGTH + 1), SEEK_CUR);
		}
	}
	if(!foundValue)
	{
		printf("%s Not Found!\n", fieldValueToCheck);
	}
	fclose(fpData);		
	pause();
}

void showAll()
{
	int successfulRead;
	char status;
	char value[LENGTH];
        FILE *fpData = fopen(DATA_FILE, "r");
    	if (!fpData) {
        	printf("Cannot open %s\n", DATA_FILE);
        	fclose(fpData);
        	return;
    	}
	while (1) {
        	for (int fieldCounter = 0; fieldCounter < fieldCount - 1; fieldCounter++) {
			if(fieldCounter == 0) {
				fseek(fpData, (fieldCount -1) * LENGTH, SEEK_CUR);
				fread(&status, sizeof(status), 1, fpData);
			}
			if(status == 'A')
			{
				if(fieldCounter == 0)
				{
					fseek(fpData, -(((fieldCount - 1)* LENGTH) + 1), SEEK_CUR);
            			}
	
				successfulRead = fread(value, LENGTH, 1, fpData);

				if (fieldCounter == fieldCount - 2)
				{
					fseek(fpData, sizeof(status), SEEK_CUR);				 
				}
				if (successfulRead == 0)
				{
					printf("End of Records\n");
					fclose(fpData);
					pause();
					return;
				}
               			printf("%s: %s\n", fieldNames[fieldCounter], value);
			}
        	}
	}	
	fclose(fpData);
}

void printMenu()
{
	printf("%s\nEnter Your Choice: ", menu);
}

void loadFields()
{
    	FILE *fpFields = fopen(FIELD_FILE, "r");
    	if (!fpFields) {
        	printf("File %s not found!\n", FIELD_FILE);
        	return;
    	}

    	char line[LENGTH];
    	fieldCount = 0;
    	while (fgets(line, sizeof(line), fpFields)) 
	{
		fieldCount++;
	}
    	rewind(fpFields);

    	fieldNames = malloc(fieldCount * sizeof(char *));
    	for (int fieldCounter = 0; fieldCounter < fieldCount; fieldCounter++) {
        	fgets(line, sizeof(line), fpFields);
        	line[strcspn(line, "\r\n")] = '\0';
		//removeNewLine(line); 
        	fieldNames[fieldCounter] = malloc(strlen(line) + 1);
	        strcpy(fieldNames[fieldCounter], line);
    	}

    	fclose(fpFields);
}
void pause()
{
	system("pause");	
}

void removeNewLine(char *text)
{
	text[strcspn(text,"\n")] = '\0';
}